src/authentication/guards/JwtAuth.guard.ts

```diff
@@ -12,21 +17,48 @@ export class JwtAuthGuard extends AuthGuard('jwt') {
   }
 
   canActivate(context: ExecutionContext) {
+    // Check if the route or class is marked as public
     const isPublic = this.reflector.getAllAndOverride(IS_PUBLIC_KEY, [
       context.getHandler(),
       context.getClass(),
     ])
 
-    // If the @Roles decorator has been used, we want to override any @PublicAccess
-    // decorator that may have been applied, and require the JWT auth
+    // Check if the route or class has specific roles specified
     const roles = this.reflector.getAllAndOverride<UserRole[]>(ROLES_KEY, [
       context.getHandler(),
       context.getClass(),
     ])
 
-    // skip JWT auth
-    if (isPublic && !roles) return true
+    // Skip JWT authentication if the route is public and does not have role restrictions
+    if (isPublic && !roles) {
+      return true
+    }
 
+    // Otherwise, require JWT authentication
     return super.canActivate(context)
   }
+
+  handleRequest(err: any, user: any, info: any) {
+    // If there's an error or the user object is missing
+    if (err || !user) {
+      // Handle invalid or expired tokens
+      if (
+        info &&
+        (info.name === 'JsonWebTokenError' || info.name === 'TokenExpiredError')
+      ) {
+        throw new HttpException(
+          {
+            statusCode: 498,
+            message: 'Invalid or expired token',
+          },
+          498,
+        )
+      }
```

Comment:
Looks like the idiomatic, recommended way to handle a custom exception, not included in the `@nest/common` package, is to create a "custom exception" that extends `HttpException`.

Might be worth looking at for this case?

https://github.com/nestjs/nest/issues/6657#issuecomment-799214959

Reply:
NestJS Custom Http Exceptions docs: https://docs.nestjs.com/exception-filters#custom-exceptions